activerecord/lib/active_record/tasks/sqlite_database_tasks.rb

```diff
@@ -30,8 +30,8 @@ def drop
       end
 
       def purge
-        drop
         connection.disconnect!
+        drop
```

Comment:
Based on the description in https://github.com/rails/rails/pull/45450, we may want to keep it around.

> sqlite3 now disconnects and reconnects the connection when purge is
called. This is necessary now that a new connection isn't created
everyt time establish_connection is called. Without this change to
purge the new database is left in an inaccessible state causing a
readonly error from the sqlite3 client. This wasn't happening in mysql
or postgres because they were already reconnecting the db connection.

I can double check, if it doesn't turn out to be a rabbit hole

Reply:
On a second look, I don't think it's weird to disconnect first actually – it might even make more logical sense since this is a sqlite adapter task and the `drop` method is deleting a file – it would definitely be more weird if this was a Postgres or MySQL task! 😄

> I'm also not following why behavior changed in this PR

The reason is that before this PR, we would never reach this codepath: https://github.com/Shopify/rails/blob/be0b5c65a175b6c92514375fc7044efb11cdbe90/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb#L38-L39

* For the primary database, we would not get there because there already exists a client (whereas after the changes, remember that we now reset `ActiveRecord::Base.connection` in the `check_protected_environments` task, so a new client must lazily be created)
* For the secondary database, previously it didn't even exist – so when `purge` calls the `drop` method, we actually raise a `NoDatabaseError` due to the database file not existing, which shortcircuits the `connection.disconnect!` call.

Reply:
I think I am going to consider this thread resolved – let me know if that's fine 🤔